Automattic/newspack-blocks
# feat(ras-acc): add GA4 events to modal checkout (redo)

### All Submissions:


### Changes proposed in this Pull Request:

Along with https://github.com/Automattic/newspack-plugin/pull/3297, reworks the GA4 events to move more of them to the front-end.

Each sequence will include the following actions:

| `action`                   | When it's used      |
| ------------------- | ------------- |
| `opened`                 | When the modal checkout is opened, either from the Donate Block, Checkout Button Block, or Variation picker screen |
| `opened_variations` | When a variation picker is opened from a Checkout button block|
| `loaded`               | When the modal completes loading |
| `continue`    | When the 'Continue' button is clicked |
|`back`            | When the 'Back' button is clicked |
|`dismissed`   | When the modal is closed before completion |
|`form_submission`   | When a submission attempt is made |
|`form_submission_success`   | When a submission attempt is completed (back-end event)

The action types used are:

| `action_type`     | When it's used      |
| ------------- | ------------- |
| `checkout_button` | When the modal trigger is the Checkout Button block |
| `donation`       | When the modal trigger is the Donation block |

We're capturing the following information on most steps:

| What's captured     | When it's captured      |
| -------------------- | --------------------- |
| `amount` | For each step, except when opening a variation picker since the price isn't known |
| `currency` | For each step |
| `is_variable` | Only the initial `open_variations` step when opening a variation picker |
| `product_id` | For each step |
| `product_type`*** | For each step |
| `recurrence` | For each step, except when opening a variation picker for a variable subscription product |
| `referer` | For each step |
| `variation_id` | For steps after picking a product variation |

*** The product types have yet to be finalized, but currently are: `product`, `subscription`, `membership`, and `donation`.

### Open questions

* ~~The "opened" action fires twice for Checkout Button blocks set to use Variable Products that the reader can select. The initial time opens the variation picker, and the second time opens the modal checkout for the product that's picked. Should there be a different `action` param for these?~~ resolved with the `opened_variations` action, added to the table above.
* When applying a coupon, the `amount` doesn't change until the form is successfully submitted (`form_submission_success`). Should this change sooner?
* When changing the price via the NYP field, the `amount` doesn't change until the form is successfully submitted (`form_submission_success`). Should this change sooner?
* ~~What `product_id` should be captured for the Donation products. Technically there's the main Donation product, plus there's a separate product for each frequency. Right now, it (incorrectly) uses both at random points.~~ -- resolved, we're grabbing the specific donation child product being purchased.

### How to test the changes in this Pull Request:

~~I've left the `console.log` in place on the GA4 `payload` so it should be relatively easy to see what's been captured and sent at each step.~~ I've removed the `console.log( payload );`, but it can be re-added under the line `window.gtag( 'event', eventName, payload );` in modal-checkout/analytics/ga4/utils/index.js.

1. Apply both this PR and https://github.com/Automattic/newspack-plugin/pull/3297, and run `npm run build` for each.
2. You may need to create products to set this up, but for testing you'll ultimately need:
* A checkout button block with a simple product
* A checkout button block with a variable product, where the reader can pick the variation
* A checkout button block with a simple subscription
* A checkout button block with a variable subscription, where the reader can pick the variation
* A checkout button block with a NYP product
* A checkout buttton block with a variable product, where the variation is pre-selected
* A checkout button block with a membership product
* A donate block.
3. Run through a checkout for each block, and track what `action`s are being fired with what information. Confirm that each step listed in the table above is represented, and is consistently populated.
4. On form submission, the `form_submission_success` is the only event fired from the back-end, so it can be tracked either by watching realtime events in Google Analytics, or by `error_log`ing the value of `$event` by adding `error_log( print_r( $event, true ) );` after [here](https://github.com/Automattic/newspack-plugin/pull/3297/files#diff-3dbb459c03feab9c03a7a641babce543932c7899efccb35734cc1cbf275633e4L154).

As a part of this PR, I also needed to add a fix similar to [this](https://github.com/Automattic/newspack-blocks/pull/1842), but for the buttons that appear in the variation picker. On epic/ras-acc, if you open the variation picker more than once on a page view, duplicate inputs were being added to the form that each button has.

When testing this PR, can you please also check:
1. When you open the variation picker from the checkout button block that a single set of hidden inputs are appended to the form, like the ones in the element inspector here: 
![CleanShot 2024-08-30 at 12 28 03](https://github.com/user-attachments/assets/b645bc4a-490e-4aee-8370-b6800e4d903a)
2. If you close and reopen the same variation picker, that no additional hidden inputs are added to each form.

### Other information:
